src/training/metrics/fbeta_multi_class_measure.py

```diff
@@ -0,0 +1,151 @@
+        # the total number of instances
+        # Shape: (num_classes, )
+        self._total_sum: Union[None, torch.Tensor] = None
+        # the total number of instances under each _predicted_ class,
+        # including true positives and false positives
+        # Shape: (num_classes, )
+        self._pred_sum: Union[None, torch.Tensor] = None
+        # the total number of instances under each _true_ class,
+        # including true positives and false negatives
+        # Shape: (num_classes, )
+        self._true_sum: Union[None, torch.Tensor] = None
+
+    @overrides
+    def __call__(
+        self,
+        predictions: torch.Tensor,
+        gold_labels: torch.Tensor,
+        mask: Optional[torch.BoolTensor] = None,
+    ):
+        """
+        # Parameters
+
+        predictions : `torch.Tensor`, required.
+            A tensor of predictions of shape (batch_size, ..., num_classes).
+        gold_labels : `torch.Tensor`, required.
+            A tensor of integer class label of shape (batch_size, ...). It must be the same
+            shape as the `predictions` tensor without the `num_classes` dimension.
+        mask : `torch.BoolTensor`, optional (default = None).
+            A masking tensor the same size as `gold_labels`.
+        """
+        predictions, gold_labels, mask = self.detach_tensors(predictions, gold_labels, mask)
+
+        # Calculate true_positive_sum, true_negative_sum, pred_sum, true_sum
+        num_classes = predictions.size(-1)
+        if (gold_labels >= num_classes).any():
+            raise ConfigurationError(
+                "A gold label passed to FBetaMeasure contains "
+                f"an id >= {num_classes}, the number of classes."
+            )
+
+        # It means we call this metric at the first time
+        # when `self._true_positive_sum` is None.
+        if self._true_positive_sum is None:
+            self._true_positive_sum = torch.zeros(num_classes, device=predictions.device)
+            self._true_sum = torch.zeros(num_classes, device=predictions.device)
+            self._pred_sum = torch.zeros(num_classes, device=predictions.device)
+            self._total_sum = torch.zeros(num_classes, device=predictions.device)
+
+        if mask is None:
+            mask = torch.ones_like(gold_labels).bool()
+        gold_labels = gold_labels.float()
+
+        threshold_predictions = torch.where(
+            predictions >= self._threshold,
+            torch.ones_like(predictions),
+            torch.zeros_like(predictions),
+        )
+        true_positives = (gold_labels == threshold_predictions) & mask
+        true_positive_sum = true_positives.sum(dim=0).float()
+        pred_sum = threshold_predictions.sum(dim=0).float()
+        true_sum = gold_labels.sum(dim=0).float()
+
+        self._true_positive_sum += true_positive_sum
+        self._pred_sum += pred_sum
+        self._true_sum += true_sum
+        self._total_sum += mask.sum().to(torch.float)
```

Comment:
Isn't this missing the calculation of averages ?

Reply:
The averaging happens in [`get_metric`](https://github.com/allenai/allennlp/blob/b3fcf60054ef8d16017f0e97c6bda82079e695db/allennlp/training/metrics/fbeta_measure.py#L165) by the looks of it (we subclass the class that implements it)

Reply:
Does it work out of the box ? I remember we had issues with it and had to re-write.

Reply:
No, as I get metrics > 100 and some of them are `NaN`. I looked at the modified `get_metric` that @nicolasahar sent but it looks virtually identical to me?

```python
	@overrides
	def get_metric(self, reset: bool = False):
		"""
		# Returns
		A tuple of the following metrics based on the accumulated count statistics:
		precisions : List[float]
		recalls : List[float]
		f1-measures : List[float]
		If `self.average` is not `None`, you will get `float` instead of `List[float]`.
		"""
		if self._true_positive_sum is None:
			raise RuntimeError("You never call this metric before.")

		tp_sum = self._true_positive_sum
		pred_sum = self._pred_sum
		true_sum = self._true_sum

		if self._average == "micro":
			tp_sum = tp_sum.sum()
			pred_sum = pred_sum.sum()
			true_sum = true_sum.sum()

		beta2 = self._beta ** 2
		# Finally, we have all our sufficient statistics.
		precision = _prf_divide(tp_sum, pred_sum)
		recall = _prf_divide(tp_sum, true_sum)
		fscore = (1 + beta2) * precision * recall / (beta2 * precision + recall)
		
		fscore[tp_sum == 0] = 0.0

		if self._average == "macro":
			precision = precision.mean()
			recall = recall.mean()
			fscore = fscore.mean()

		if reset:
			self.reset()

		if self._average is None:
			return {
				"precision": precision.tolist(),
				"recall": recall.tolist(),
				"fscore": fscore.tolist(),
			}
		else:
			return {"precision": precision.item(), "recall": recall.item(), "fscore": fscore.item()}
```

In any case, I tried this in place of the [`get_metric`](https://github.com/allenai/allennlp/blob/b3fcf60054ef8d16017f0e97c6bda82079e695db/allennlp/training/metrics/fbeta_measure.py#L165) in the parent class and ran into the same issues.

Reply:
Let me know if you find out what is going on, I think that what Nick sent you was tested and seemed to give correct numbers so maybe the bug is somewhere earlier in the pipeline ?

Reply:
Do you know what version of AllenNLP the code I was sent was tested against?

Reply:
I believe 0.9.1 (unreleased). Did you end up figuring this out or want to jump on a debug call?

Reply:
Hmm okay, I am using 1.0 pre-release. Haven't figured it out yet, let me spend some more time on it. Are you able to get me the exact dataset you used in your setup where this worked? Along with the performance scores you got? This would help me figure out if its a dataset issue or model issue.